### ISSUE TITLE
`yield` is missing in `no-unsafe-finally`

### ISSUE THREAD

**Issue by magic-akari** [eslint/eslint#14806]
**What rule do you want to change?**
no-unsafe-finally

**Does this change cause the rule to produce more or fewer warnings?**
It will produce more warnings.

**How will the change be implemented? (New option, new default behavior, etc.)?**
new default behavior

**Please provide some example code that this change will affect:**

```js
/*eslint no-unsafe-finally: "error"*/
let foo = function*() {
    try {
        return 1;
    } catch(err) {
        return 2;
    } finally {
        yield 3;
    }
};
```

**What does the rule currently do for this code?**
No warnings.

**What will the rule do after it's changed?**
It will report warning with message `Unsafe usage of YieldExpression`.

**Are you willing to submit a pull request to implement this change?**
Sure, I do. I will submit a pull request.

**Comment by hax** [comment]
I'd like to provide an example of how `yield` in `finally` could cause problem:

```js
function *g() {
  try {
    console.log('do sth')
  } finally {
    yield
    console.log('clean up')
  }
}

for (let x of g()) break
```

In this simple example, the code only "do sth" but not "clean up" (the `return()` method of the generator will produce `{done: false}`.)

**Comment by nzakas** [comment]
I'm not sure if `yield` is unsafe in this situation. Unlike `break`, `continue`, `return`, and `throw`, `yield` doesn't actually exit the normal control flow of the function -- it just pauses and then picks up again after that. I'm not sure @hax's example is a problem, because it is the `break` statement that is preventing the second `console.log()` from firing.

That said, this is a bit of an edge case, so it's possible I'm not thinking about this clearly. I'd like to hear what others think.

**Comment by nzakas** [comment]
@aladdin-add please be sure to leave a comment explaining your position and what you'd like feedback on when moving an issue to "needs feedback". Thanks.

**Comment by nzakas** [comment]
**TSC Summary:** This proposal seeks to add a check for `yield` into `no-unsafe-finally`

**TSC Question:** shall we accept this proposal?

**Comment by btmills** [comment]
Mentally diagramming the control flow for these cases took a bit, but I ended up in the same place as @nzakas: `yield` inside `finally` is different from the other control flow statements that `no-unsafe-finally` detects because `yield` can resume normal control flow. While I've never had a use case for it, I can imagine a safe use case, perhaps to return a termination sentinel value for example.

If you still want to forbid `yield` inside `finally`, you can use `no-restricted-syntax` ([demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLXJlc3RyaWN0ZWQtc3ludGF4OiBbXCJlcnJvclwiLCBcIlRyeVN0YXRlbWVudCA+IC5maW5hbGl6ZXIgWWllbGRFeHByZXNzaW9uXCJdICovXG5cbmZ1bmN0aW9uKiB0ZXN0KCkge1xuICB5aWVsZCAnYWxsb3dlZCc7XG4gIHRyeSB7XG4gICAgeWllbGQgJ2FsbG93ZWQnO1xuICB9IGZpbmFsbHkge1xuICAgIHlpZWxkICdmb3JiaWRkZW4nO1xuICB9XG59Iiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjoxMiwic291cmNlVHlwZSI6Im1vZHVsZSIsImVjbWFGZWF0dXJlcyI6eyJqc3giOnRydWV9fSwicnVsZXMiOnt9LCJlbnYiOnt9fX0=)):

```js
/* eslint no-restricted-syntax: ["error", "TryStatement > .finalizer YieldExpression"] */

function* test() {
  yield 'allowed';
  try {
    yield 'allowed';
  } finally {
    yield 'forbidden';
  }
}
```

**Comment by hax** [comment]
I hope this issue could be reconsider.

The example code shows the problem, if u put `yield` in finally it actually make `finally` block not work, and may cause memory leak. It obviously "unsafe" in most cases.

Theoretically someone may really want the effect of that, but like @nzakas said it's a edge case, I don't know anyone write such code intentionally, so just like `return` or other flow control statements in `finally` block, it should be forbidden by default.
